Transaction queue

ablax opened the pull request:
- [x] Create transaction pool
- [x] Create transaction queue
- [x] Prepare for transaction validation

bokoto000 committed:
Transaction queue and pool

Commit:
refactor: rename packages to fix checkstyle error

Commit:
feat: pop transaction with timer

Commit:
fix: build, change constructor access

Commit:
refactor: replace equals with lombok

Commit:
chore: delete obsolete file

Commit:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

# Conflicts:
#	src/main/java/com/limechain/network/protocol/transaction/TransactionsEngine.java

Commit:
chore: prepare for transaction validaiton

Commit:
chore: remove stacktrace

Commit:
chore: remove redundant return

Commit:
chore: remove stacktrace

Commit:
rethrow interrupted exception

Commit:
refactor: limit constructor access

bokoto000 commented on src/main/java/com/limechain/network/protocol/transaction/TransactionsEngine.java:
```diff
@@ -126,7 +142,9 @@ public void writeHandshakeToStream(Stream stream, PeerId peerId) {
         try (ScaleCodecWriter writer = new ScaleCodecWriter(buf)) {
-            writer.write(new TransactionsWriter(), new byte[][]{new byte[]{}, new byte[]{}});
+            writer.write(new TransactionsWriter(), new Extrinsics[]{
+                    new Extrinsics(new byte[]{}), new Extrinsics(new byte[]{})
```
Add TODO to replace empty transaction messages in future

bokoto000 commented on src/main/java/com/limechain/network/protocol/transaction/state/TransactionState.java:
```diff
@@ -0,0 +1,103 @@
+            ValidTransaction transaction = null;
+            while (transaction == null) {
+                transaction = popTransaction();
+                Thread.sleep(50);
```
Why do we use sleep here

ablax replied:
To avoid high CPU usage in this almost infinite loop

bokoto000 commented on src/main/java/com/limechain/network/protocol/transaction/state/TransactionState.java:
```diff
@@ -0,0 +1,103 @@
+        return transactionPool.insert(validTransaction);
+    }
+
+    //public void notifyStatus
```
Add as TODO

bokoto000 commented on src/main/java/com/limechain/network/protocol/transaction/state/ValidTransaction.java:
```diff
@@ -0,0 +1,37 @@
+        return new ValidTransactionComparator().compare(this, transaction);
+    }
+
+    static class ValidTransactionComparator implements Comparator<ValidTransaction> {
```
Maybe rename this to ValidTransactionPriorityComparator as I'm not sure if its clear that we are comparing by priority

Commit:
chore: remove redundant comparator

georg-getz commented on src/main/java/com/limechain/network/protocol/transaction/state/Pool.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@NoArgsConstructor
```
No arg constructor is implied if no other constructor is found.

georg-getz commented on src/main/java/com/limechain/network/protocol/transaction/state/ValidTransaction.java:
```diff
@@ -0,0 +1,29 @@
+    @Getter
+    private Validity validity;
+
+    public ValidTransaction(byte[] extrinsic){
+        this.extrinsic = extrinsic;
+    }
+
+    public ValidTransaction(byte[] extrinsic, Validity validity){
+        this.extrinsic = extrinsic;
+        this.validity = validity;
+    }
```
space before opening curly brace `{`

georg-getz commented on src/main/java/com/limechain/network/protocol/transaction/state/Validity.java:
```diff
@@ -0,0 +1,24 @@
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class Validity {
+    private BigInteger priority;
+    private byte[][] requires;
+    private byte[][] provides;
+    private BigInteger longevity;
+    private boolean propagate;
+
+    public Validity(BigInteger priority){
+        this.priority = priority;
+    }
+}
```
Add a comment explaining what this class represents and how the priority works. If I have priority 1 and someone else has 18, am I with a higher priority or not? Also how are the negative cases handled? Would be nice to have a test for that case as well.

Commit:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

# Conflicts:
#	src/main/java/com/limechain/network/protocol/transaction/TransactionsEngine.java

Commit:
chore: fix imports

Commit:
chore: reformat

ablax commented:
Currently blocked, as it depends on Trie

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

# Conflicts:
#	src/main/java/com/limechain/network/protocol/transaction/TransactionsEngine.java

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

Zurcusa committed:
feat: Implement transaction validation.

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

# Conflicts:
#	src/main/java/com/limechain/rpc/server/RpcApp.java
#	src/main/java/com/limechain/runtime/RuntimeEndpoint.java

Zurcusa committed:
feat: Implement transaction queue and pool population logic.

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

Zurcusa committed:
refactor: Move block announce handshake at start of full sync. Handshake all peers at end of full sync.

Zurcusa committed:
chore: Address some of the comments.

Zurcusa committed:
fix: reverse priority order so that it follows a reversed natural order.

Grigorov-Georgi commented on src/main/java/com/limechain/transaction/TransactionState.java:
```diff
@@ -0,0 +1,119 @@
+        transactionQueue.add(validTransaction);
+    }
+
+    public ValidTransaction popTransaction() {
```
Not a big deal, but since 'pop' semantic suggests a stack-related operation, it might be better to rename it to pollTransaction for clarity.

Zurcusa committed:
chore: add todo

Zurcusa committed:
chore: Address comments.

georg-getz commented on src/main/java/com/limechain/transaction/TransactionPool.java:
```diff
@@ -0,0 +1,46 @@
+
+public class TransactionPool {
+
+    private final Map<byte[], ValidTransaction> transactions;
+
+    public TransactionPool() {
+        transactions = new HashMap<>();
+    }
```
You can move the initialization of the map to the field declaration and delete the constructor entirely

Zurcusa replied:
I usually prefer keeping init code inside of the ctors, especially in spring projects. Since in spring ctor bean injection is preferred it helps keep a uniform init standard in the whole project.

georg-getz commented on src/main/java/com/limechain/transaction/TransactionState.java:
```diff
@@ -0,0 +1,119 @@
+            log.severe("Error while waiting for transaction: " + e.getMessage());
+        } catch (TimeoutException e) {
+            futureTransaction.cancel(true);
+        } catch (InterruptedException e) {
+            if (Thread.interrupted())
+                throw new InterruptedException();
+        }
```
InterruptedException is a checked excaption and we don't want to throw it, we have ThreadInterruptedException for cases like this one

georg-getz commented on src/test/java/com/limechain/network/protocol/transaction/state/TransactionStateTest.java:
```diff
@@ -0,0 +1,126 @@
+    private static final byte[] TAG_1 = "tag1".getBytes();
+    private static final byte[] TAG_2 = "tag2".getBytes();
+
+    private TransactionState sut;
```
can you rename this field

Zurcusa replied:
I always name the class I test following this approach https://en.wikipedia.org/wiki/System_under_test

Zurcusa committed:
chore: Address comments.

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

georg-getz commented on src/main/java/com/limechain/transaction/TransactionState.java:
```diff
@@ -0,0 +1,120 @@
+        } catch (TimeoutException e) {
+            futureTransaction.cancel(true);
+        } catch (InterruptedException e) {
+            if (Thread.interrupted())
+                throw new ThreadInterruptedException(e);
```
If you are catching an `InterruptedException` then you don't need to check whether the thread was interrupted but rather force it to be interrupted like this `Thread.currentThread().interrupt();`
This will clear the sonar cloud error

Zurcusa committed:
chore: Address comments.

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

Zurcusa committed:
feat: Keep up with the head of the chain. (#591)

# Description

- What does this PR do?
Our node needs to keep up with the head of the chain. In other words,
all announced blocks must be executed, and their states must be kept in
memory before being finalized. This PR implements the functionality to
do so. It also exposes a runtime API interface. This makes it more
readable and easily extensible. A few other changes were made along the
way that improved performance (async calls) and structured the project
in a better way.

- Why are these changes needed?
We need the latest state of the chain to validate transactions.

Fixes #574 and #573 

## Checklist:
- [X] I have read the [contributing
guidelines](https://github.com/LimeChain/Fruzhin/blob/dev/CONTRIBUTING.md).
- [X] My PR title matches the [Conventional Commits
spec](https://www.conventionalcommits.org/).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

Zurcusa committed:
chore: Merge dev. Solve conflicts.

Zurcusa committed:
refactor: Encapsulate TrieAccessor into the Runtime instance by removing Runtime.getTrieAccessor().

Zurcusa committed:
Merge remote-tracking branch 'origin' into #201-transaction-queue

Zurcusa committed:
fix: Use DigestHelper to properly assign primary blocks in the block tree.

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

Zurcusa committed:
refactor: Static lock transaction engine to avoid the same transaction being validated twice.

Zurcusa committed:
Merge remote-tracking branch 'origin/dev' into #201-transaction-queue

# Conflicts:
#	src/main/java/com/limechain/rpc/config/CommonConfig.java
#	src/main/java/com/limechain/runtime/Runtime.java
#	src/main/java/com/limechain/sync/fullsync/FullSyncMachine.java

Zurcusa committed:
chore: Merge dev.

Zurcusa committed:
feat: Revert state storage updates for runtime calls... (#602)

# Description
Runtime now reverts all state storage changes originating from a runtime
call different from `Core_execute_block`. This is needed when validating
transactions. The runtime updates state storage when checking if a
provided transaction is valid.

Zurcusa committed:
Merge remote-tracking branch 'origin/#201-transaction-queue' into #201-transaction-queue

# Conflicts:
#	src/main/java/com/limechain/runtime/RuntimeImpl.java

Zurcusa committed:
refactor: Stabilize the block import + transaction validation process.
